notify-debouncer-mini/src/lib.rs

```diff
@@ -160,88 +158,126 @@ pub struct DebouncedEvent {
+    oldest_update: Instant,
 }
 
 impl DebounceDataInner {
+    pub fn new(timeout: Duration) -> Self {
+        let time = Instant::now();
+        Self {
+            timeout: timeout,
+            oldest_update: time,
+            last_run: time,
+            d: Default::default(),
+        }
+    }
+
+    /// Returns a duration to wait for the next tick
+    #[inline]
+    pub fn next_tick(&self) -> Option<Duration> {
+        if self.has_tick_data() {
+            let next_run = self.oldest_update.checked_add(self.timeout)
+                .map_or(self.timeout, |v|v.saturating_duration_since(Instant::now()));
+            Some(next_run)
+        } else {
+            None
+        }
+    }
+
+    /// Returns true if there are buffered events
+    #[inline]
+    pub fn has_tick_data(&self) -> bool {
+        !self.d.is_empty()
+    }
+
     /// Retrieve a vec of debounced events, removing them if not continuous
+    ///
+    /// Updates the internal tracker for the next tick
     pub fn debounced_events(&mut self) -> Vec<DebouncedEvent> {
         let mut events_expired = Vec::with_capacity(self.d.len());
         let mut data_back = HashMap::with_capacity(self.d.len());
         // TODO: perfect fit for drain_filter https://github.com/rust-lang/rust/issues/59618
+        // reset oldest update
+        self.oldest_update = Instant::now();
         for (k, v) in self.d.drain() {
             if v.update.elapsed() >= self.timeout {
                 events_expired.push(DebouncedEvent::new(k, DebouncedEventKind::Any));
             } else if v.insert.elapsed() >= self.timeout {
                 data_back.insert(k.clone(), v);
                 events_expired.push(DebouncedEvent::new(k, DebouncedEventKind::AnyContinuous));
             } else {
+                if self.oldest_update > v.update {
+                    self.oldest_update = v.update;
+                }
                 data_back.insert(k, v);
             }
         }
         self.d = data_back;
+        self.last_run = Instant::now();
         events_expired
     }
 
-    /// Returns all currently stored errors
-    pub fn errors(&mut self) -> Vec<Error> {
-        let mut v = Vec::new();
-        std::mem::swap(&mut v, &mut self.e);
-        v
-    }
-
-    /// Add an error entry to re-send later on
-    pub fn add_error(&mut self, e: crate::Error) {
-        self.e.push(e);
-    }
-
     /// Add new event to debouncer cache
-    pub fn add_event(&mut self, e: Event) {
+    #[inline(always)]
+    fn add_event(&mut self, e: Event) {
+        let time = Instant::now();
+        if self.d.is_empty() {
+            self.oldest_update = time;
+        }
         for path in e.paths.into_iter() {
             if let Some(v) = self.d.get_mut(&path) {
-                v.update = Instant::now();
+                v.update = time;
             } else {
-                self.d.insert(path, EventData::new_any());
+                self.d.insert(path, EventData::new_any(time));
```

Comment:
Pretty same comment about the one character variables. We're not writing `C` in the end and have plenty of space to write proper descriptive names.